File: Contents/InitialConsonantSearch/InitialConsonantSearch/Main/IndicatorSettingSearchController.m
//
//  IndicatorSettingSearchController.m
//  IndicatorSettingsTest
//
//  Created by Kwan Hyun Son on 10/18/23.
//

#import <BaseKit/BaseKit.h>
#import "IndicatorSettingSearchController.h"
#import "IndicatorSettingCell.h"

@interface IndicatorSettingSearchController () <UITableViewDataSource, UITableViewDelegate, UIGestureRecognizerDelegate>
@property (weak, nonatomic) IBOutlet UITableView *tableView;
@property (nonatomic, strong) MGUNOSearchResultView *noSearchResultView;
@property (nonatomic, strong) UITableViewDiffableDataSource <NSString *, MGROutlineItem <DTOIndicatorSetting *>*>*dataSource;
@property (nonatomic, copy) MGRDispatchDebounceBlock debounceBlock;

@property (nonatomic, assign) BOOL showKeyboard;
@property (nonatomic, strong) id <NSObject> showObserver;
@property (nonatomic, strong) id <NSObject> hideObserver;
@property (nonatomic, strong, nullable) NSNotificationName previousNotificationName;
@property (nonatomic, assign) CGFloat previousKeyboardHeight;
@end


@implementation IndicatorSettingSearchController

#pragma mark - Override

- (void)dealloc {
    [[NSNotificationCenter defaultCenter] removeObserver:_showObserver];
    [[NSNotificationCenter defaultCenter] removeObserver:_hideObserver];
}

- (instancetype)initWithNibName:(NSString *)nibNameOrNil bundle:(NSBundle *)nibBundleOrNil {
    self = [super initWithNibName:nibNameOrNil bundle:nibBundleOrNil];
    if (self) {
        CommonInit(self);
    }
    return self;
}

- (void)awakeFromNib {
    [super awakeFromNib];
    CommonInit(self);
}

- (void)viewDidLoad {
    [super viewDidLoad];
    _debounceBlock = MGRDispatchDebounceMake();
    self.view.backgroundColor = [UIColor clearColor];
    UIView *backgroundView = [UIView new];
    backgroundView.backgroundColor = [UIColor clearColor];
    
    // 여백 터치 및 팬 제스처 시에 내려버린다
    UITapGestureRecognizer *tapGestureRecognizer = [[UITapGestureRecognizer alloc] initWithTarget:self action:@selector(endSearchMode:)];
    tapGestureRecognizer.numberOfTapsRequired = 1;
    tapGestureRecognizer.delegate = self;
    [self.view addGestureRecognizer:tapGestureRecognizer];
    
    UIPanGestureRecognizer *panGestureRecognizer = [[UIPanGestureRecognizer alloc] initWithTarget:self action:@selector(endSearchMode:)];
    panGestureRecognizer.maximumNumberOfTouches = 1;
    panGestureRecognizer.delegate = self;
    [self.view addGestureRecognizer:panGestureRecognizer];
    
    _noSearchResultView = [MGUNOSearchResultView new];
    self.noSearchResultView.imageType = MGUNOSearchResultViewImageTypeDynamic;
    self.noSearchResultView.hidden = YES;
    
    [backgroundView addSubview:self.noSearchResultView];
    [self.noSearchResultView mgrPinHorizontalEdgesToSuperviewEdges];
    [self.noSearchResultView.topAnchor constraintEqualToAnchor:backgroundView.topAnchor constant:20.0].active = YES;
    self.tableView.backgroundView = backgroundView;
    
    /// 디퍼블을 이용하느냐 마느냐.
    [self configureDataSource];
    [self setupKeyboardObserver];
}

- (void)viewWillTransitionToSize:(CGSize)size
       withTransitionCoordinator:(id<UIViewControllerTransitionCoordinator>)coordinator {
    [super viewWillTransitionToSize:size withTransitionCoordinator:coordinator];
    [coordinator animateAlongsideTransition:^(id<UIViewControllerTransitionCoordinatorContext>  _Nonnull context) {
    } completion:^(id<UIViewControllerTransitionCoordinatorContext>  _Nonnull context) { 
    }];
    if (self.showKeyboard == YES) {
        [self.view endEditing:YES];
    }
}

#pragma mark - 생성 & 소멸

static void CommonInit(IndicatorSettingSearchController *self) {
    self->_debounceEnabled = YES;
}

- (void)configureDataSource {
    self.tableView.dataSource = nil;
    
    __weak __typeof(self) weakSelf = self;
    _dataSource =
    [[UITableViewDiffableDataSource alloc] initWithTableView:self.tableView
                                                cellProvider:^UITableViewCell *(UITableView *tableView,
                                                                                NSIndexPath *indexPath,
                                                                                MGROutlineItem <DTOIndicatorSetting *>*outlineItem) {
        __strong __typeof(weakSelf) self = weakSelf;
        NSString *identifier = IndicatorSettingCellIDSearch;
        
        IndicatorSettingCell *cell =
        [tableView dequeueReusableCellWithIdentifier:identifier
                                        forIndexPath:indexPath];
        cell.selectionStyle = UITableViewCellSelectionStyleNone;
        if ([cell isKindOfClass:[IndicatorSettingCell class]] == NO) {
            NSAssert(FALSE, @"Could not create new cell");
        }
        
        cell.data = outlineItem.contentItem;
        return cell;
    }];

    self.dataSource.defaultRowAnimation = UITableViewRowAnimationFade;
    // 없는 것이 맞다
    // [self.dataSource applySnapshot:self.viewModel.snapshotForCurrentState animatingDifferences:NO];
}

- (void)setupKeyboardObserver {
    NSNotificationCenter *nc = [NSNotificationCenter defaultCenter]; // search filterView의 frame을 이동시키기 위해 사용한다.
    __weak __typeof(self)weakSelf = self;
    self.showObserver = [nc addObserverForName:UIKeyboardWillShowNotification
                                        object:nil
                                         queue:[NSOperationQueue mainQueue]
                                    usingBlock:^(NSNotification * _Nonnull note) {
        weakSelf.showKeyboard = YES;
        [weakSelf handleKeyboardNotification:note];
    }];
    self.hideObserver = [nc addObserverForName:UIKeyboardWillHideNotification // 키보드가 내려갈 때
                                        object:nil
                                         queue:[NSOperationQueue mainQueue]
                                    usingBlock:^(NSNotification * _Nonnull note) {
        weakSelf.showKeyboard = NO;
        [weakSelf handleKeyboardNotification:note];
    }];
}

#pragma mark - 세터 & 게터

- (void)setEmptyStringMode:(BOOL)emptyStringMode {
    _emptyStringMode = emptyStringMode;
    if (emptyStringMode == YES) { // 문자열이 없다면 (nil or empty)
        if (self.tableView.alpha != 0.0) {
            self.tableView.alpha = 0.0;
            self.noSearchResultView.hidden = YES; // debounce 때문에 여기서만 hidden 처리
            [self.dataSource applySnapshot:[NSDiffableDataSourceSnapshot new] // debounce 때문에 바로 처리
                      animatingDifferences:NO];
            self.debounceBlock(0.0, ^{}); // 기존 디바운스 예약을 무력화 시킨다
        }
    } else {
        if (self.tableView.alpha != 1.0) {
            self.tableView.alpha = 1.0;
            // self.noSearchResultView.hidden = NO; // debounce 때문에 여기서는 이 라인을 실행해서는 안된다
        }
    }
}

#pragma mark - Actions

- (void)handleKeyboardNotification:(nullable NSNotification *)notification {
    NSNotificationName previousNotificationName = self.previousNotificationName;
    CGFloat previousKeyboardHeight = self.previousKeyboardHeight;
    NSNotificationName name = notification.name; // UIKeyboardWillShowNotification OR UIKeyboardWillHideNotification
    self.previousNotificationName = name;
    
    NSValue *keyboardFrame = notification.userInfo[UIKeyboardFrameEndUserInfoKey];
    if (notification.userInfo == nil || keyboardFrame == nil) {
        return;
    }
    CGFloat keyboardHeight = keyboardFrame.CGRectValue.size.height;
    self.previousKeyboardHeight = keyboardHeight;
    if ([previousNotificationName isEqualToString:name] && previousKeyboardHeight == keyboardHeight) {
        return;
    }
    
    NSNumber *durationNumber = notification.userInfo[UIKeyboardAnimationDurationUserInfoKey];
    NSNumber *animationCurveNumber = notification.userInfo[UIKeyboardAnimationCurveUserInfoKey]; // 설명은 아래
    // NSNumber *isLocalUserNumber = notification.userInfo[UIKeyboardIsLocalUserInfoKey]; // NSNumber of BOOL
    
    NSTimeInterval duration = durationNumber.doubleValue;
    UIViewAnimationOptions animationCurve = [animationCurveNumber unsignedIntegerValue];
    // BOOL isLocalUser = [isLocalUserNumber boolValue];
    
    /// Hide
    if ([name isEqualToString:UIKeyboardWillHideNotification]) {
        CGFloat originalMaxOffsetY = self.tableView.mgrMaxOffset.y - self.tableView.contentInset.bottom;
        if (originalMaxOffsetY < self.tableView.contentOffset.y) {
            [UIViewPropertyAnimator runningPropertyAnimatorWithDuration:duration delay:0.0 options:animationCurve animations:^{
                [self.tableView setContentOffset:CGPointMake(0.0, originalMaxOffsetY)];
                [self.view layoutIfNeeded];
            } completion:^(UIViewAnimatingPosition finalPosition) {
                self.tableView.contentInset = UIEdgeInsetsZero;
                self.tableView.scrollIndicatorInsets = UIEdgeInsetsZero;
            }];
        } else {
            self.tableView.contentInset = UIEdgeInsetsZero;
            self.tableView.scrollIndicatorInsets = UIEdgeInsetsZero;
        }
        return; //! return 해야한다.
    }
    
    /// Show
    if ([name isEqualToString:UIKeyboardWillShowNotification]) {
        CGFloat bottomIset = keyboardHeight - self.view.safeAreaInsets.bottom;
        self.tableView.contentInset = UIEdgeInsetsMake(0.0, 0.0, bottomIset, 0.0);
        self.tableView.scrollIndicatorInsets = UIEdgeInsetsMake(0.0, 0.0, bottomIset, 0.0);
    }

    UIWindow *window = self.view.window;
    CGRect rect = [self.view convertRect:self.view.bounds toView:window];
    
    CGFloat upLength = rect.origin.y + rect.size.height + 4.0; // 4.0은 마진..
    CGFloat total = upLength + keyboardHeight;
    CGPoint movingOffset = self.tableView.contentOffset;
    if (total > window.bounds.size.height) {
        CGFloat move = ABS(window.bounds.size.height - total);
        movingOffset = CGPointMake(movingOffset.x, movingOffset.y + move);
    }
    [UIViewPropertyAnimator runningPropertyAnimatorWithDuration:duration delay:0.0 options:animationCurve animations:^{
        [self.tableView setContentOffset:movingOffset];
        [self.view layoutIfNeeded];
    } completion:^(UIViewAnimatingPosition finalPosition) {
    }];
}

#pragma mark - <UITableViewDataSource>

- (NSInteger)numberOfSectionsInTableView:(UITableView *)tableView {
    return 0;
}

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section {
    return 0;
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath {
    return nil;
}

#pragma mark - <UITableViewDelegate>

- (void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath {
    MGROutlineItem <DTOIndicatorSetting *>*outlineItem = [self.dataSource itemIdentifierForIndexPath:indexPath];
    if (self.selectItemCompletion != nil) {
        self.selectItemCompletion(outlineItem);
    }
}

#pragma mark - <UISearchBarDelegate> 모두 Optional 전달 받아서 처리한다

- (BOOL)searchBarShouldBeginEditing:(UISearchBar *)searchBar {
    [self updateEmptyStringModeForText:searchBar.text];
    return YES;
}

- (void)searchBar:(UISearchBar *)searchBar textDidChange:(NSString *)searchText {
    [self updateEmptyStringModeForText:searchText];
    if (self.emptyStringMode == NO) {
        if (self.debounceEnabled == YES) {
            __weak __typeof(self) weakSelf = self;
            self.debounceBlock(0.2, ^{
                [weakSelf updateSearchResult:searchBar currentText:searchText];
            });
        } else {
            [self updateSearchResult:searchBar currentText:searchText];
        }
    }
}

#pragma mark - <UIGestureRecognizerDelegate>

- (BOOL)gestureRecognizerShouldBegin:(UIGestureRecognizer *)gestureRecognizer {
    CGPoint location = [gestureRecognizer locationInView:self.tableView];
    if ([self.tableView indexPathForRowAtPoint:location] != nil && self.emptyStringMode == NO) {
        return NO;
    } else {
        return YES;
    }
}

- (void)endSearchMode:(UIGestureRecognizer *)sender {
    [self.view.window endEditing:YES];
}

#pragma mark - Private Helper

- (void)updateEmptyStringModeForText:(NSString *)searchText {
    if (searchText == nil || searchText.length == 0) {
        self.emptyStringMode = YES;
    } else {
        self.emptyStringMode = NO;
    }
}

/// 디바운스(debounce) 효과가 부담스럽다면 디바운스만 제거하고 실행 Block를 그냥 실행하면된다
- (void)updateSearchResult:(UISearchBar *)searchBar currentText:(NSString *)currentText {
    NSDiffableDataSourceSnapshot<NSString *, MGROutlineItem <DTOIndicatorSetting *>*>*snp = [self.viewModel snapshotForCurrentText:currentText];
    if (snp.numberOfItems == 0) {
        self.noSearchResultView.searchText = currentText;
        self.noSearchResultView.hidden = NO;
        [self.dataSource applySnapshot:[NSDiffableDataSourceSnapshot new] animatingDifferences:NO];
    } else {
        self.noSearchResultView.hidden = YES;
        self.tableView.contentOffset = CGPointZero;
        [self.dataSource applySnapshot:snp animatingDifferences:NO];
    }
}

@end

//    NSDiffableDataSourceSnapshot <NSNumber *, MGPUser *>*snapshot  = [NSDiffableDataSourceSnapshot new];
//    [snapshot appendSectionsWithIdentifiers:@[@(RecipeListSectionMain)]];
//    if (searchText == nil || [searchText isEqualToString:@""] == YES) {
//
//        [snapshot appendItemsWithIdentifiers:self.users intoSectionWithIdentifier:@(RecipeListSectionMain)];
//
//    } else {
//        NSArray <MGPUser *>*result = [self.users mgrFilter:^BOOL(MGPUser *user) {
//            return [user.name containsString:searchText];
//        }];
//        [snapshot appendItemsWithIdentifiers:result intoSectionWithIdentifier:@(RecipeListSectionMain)];
//    }
//    [self.dataSource applySnapshot:snapshot animatingDifferences:YES completion:^{}];

//- (void)performQueryWith:(NSString * _Nullable)filter {
//    
//    NSArray <Mountain *>*mountains = [self.mountainsController filteredMountainsWith:filter
//                                                                               limit:NSNotFound];
//    mountains = [mountains sortedArrayUsingComparator:^NSComparisonResult(Mountain *obj1, Mountain *obj2) {
//        return [[obj1 name] compare:[obj2 name]];
//    }];
//    
//    NSDiffableDataSourceSnapshot <MGRMainSection, Mountain *>*snapshot = [NSDiffableDataSourceSnapshot new];
//    [snapshot appendSectionsWithIdentifiers:@[mainSection]];
//    [snapshot appendItemsWithIdentifiers:mountains];
//    [self.dataSource applySnapshot:snapshot animatingDifferences:YES];
//}

//- (NSArray <Mountain *>*)filteredMountainsWith:(NSString * _Nullable)filter
//                                         limit:(NSInteger)limit {
//    
//    NSArray <Mountain *>*mountains = self.mountains;
//    NSMutableArray <Mountain *>*filtered = [NSMutableArray array];
//    for (Mountain *mountain in mountains) {
//        if ([mountain contains:filter] == YES) {
//            [filtered addObject:mountain];
//        }
//    }
//    
//    if (limit != NSNotFound) {
//        return [filtered mgrSubArrayToIndex:limit]; // ex : subArrayToIndex:2    // 0, 1, 2
//    } else {
//        return filtered;
//    }
//}

//- (BOOL)contains:(NSString * _Nullable)filter {
//    if (filter == nil) {
//        return YES;
//    } else if ([filter isEqualToString:@""] == YES) {
//        return YES;
//    }
//    
//    NSString *lowercasedFilter = [filter lowercaseString];
//    NSString *lowercasedName = [self.name lowercaseString];
//    return [lowercasedName containsString:lowercasedFilter];
//}

File: Contents/InitialConsonantSearch/InitialConsonantSearch_Swift/Main/IndicatorSettingSearchController.swift
//
//  IndicatorSettingSearchController.swift
//  IndicatorSettingsTest_Swift
//
//  Created by Kwan Hyun Son on 10/18/23.
//

import UIKit
import BaseKit
import IosKit

final class IndicatorSettingSearchController: UIViewController {
    
    // MARK: - Property
    
    var viewModel: IndicatorSettingViewModel?
    var emptyStringMode = true {
        didSet {
            if emptyStringMode {
                if tableView.alpha != 0.0 {
                    tableView.alpha = 0.0
                    noSearchResultView.isHidden = true // debounce 때문에 여기서만 hidden 처리
                    dataSource?.apply( // debounce 때문에 바로 처리
                        NSDiffableDataSourceSnapshot(),
                        animatingDifferences: false)
                    debounceClosure(0.0) {} // 기존 디바운스 예약을 무력화 시킨다
                }
            } else {
                if tableView.alpha != 1.0 {
                    tableView.alpha = 1.0
                    // self.noSearchResultView.hidden = NO; // debounce 때문에 여기서는 이 라인을 실행해서는 안된다
                }
            }
        }
    }
    var selectItemCompletion: ((SKHOutlineItem<DTOIndicatorSetting>) -> Void)?
    @IBOutlet weak var tableView: UITableView!
    var noSearchResultView: SKUNOSearchResultView = SKUNOSearchResultView()
    var dataSource: UITableViewDiffableDataSource<String, SKHOutlineItem<DTOIndicatorSetting>>?
    var debounceClosure = SKHDispatchDebounceMake()
    var isShowKeyboard = false
    var showObserver: NSObjectProtocol?
    var hideObserver: NSObjectProtocol?
    var previousNotificationName: NSNotification.Name?
    var previousKeyboardHeight: CGFloat = 0
    var isDebounceEnabled = true // Debounce 효과를 없애고 싶다면 false 하라. IndicatorSettingViewController에서 false 하라.
    
    // MARK: - Override
    
    deinit {
        if let showObserver = showObserver {
            NotificationCenter.default.removeObserver(showObserver)
        }
        if let hideObserver = showObserver {
            NotificationCenter.default.removeObserver(hideObserver)
        }
    }
    
    override func viewDidLoad() {
        super.viewDidLoad()
        view.backgroundColor = .clear
        let backgroundView = UIView()
        backgroundView.backgroundColor = .clear
        
        let tapGestureRecognizer = UITapGestureRecognizer(target: self, action: #selector(endSearchMode(_:)))
        tapGestureRecognizer.numberOfTapsRequired = 1
        tapGestureRecognizer.delegate = self
        view.addGestureRecognizer(tapGestureRecognizer)
        
        let panGestureRecognizer = UIPanGestureRecognizer(target: self, action: #selector(endSearchMode(_:)))
        panGestureRecognizer.maximumNumberOfTouches = 1
        panGestureRecognizer.delegate = self
        view.addGestureRecognizer(panGestureRecognizer)
        
        noSearchResultView.imageType = .dynamic
        noSearchResultView.isHidden = true
        
        backgroundView.addSubview(noSearchResultView)
        noSearchResultView.skhPinHorizontalEdgesToSuperviewEdges()
        noSearchResultView.topAnchor.constraint(equalTo: backgroundView.topAnchor, constant: 20.0).isActive = true
        tableView.backgroundView = backgroundView
        
        // 디퍼블을 이용하느냐 마느냐.
        configureDataSource()
        setupKeyboardObserver()
    }
    
    override func viewWillTransition(to size: CGSize, with coordinator: UIViewControllerTransitionCoordinator) {
        super.viewWillTransition(to: size, with: coordinator)
        coordinator.animate { context in
        } completion: { context in
        }
        if isShowKeyboard {
            view.endEditing(true)
        }
    }
    
    // MARK: - 생성 & 소멸
    
    func configureDataSource() {
        tableView.dataSource = nil
        dataSource = UITableViewDiffableDataSource<String, SKHOutlineItem<DTOIndicatorSetting>>(
            tableView: tableView) { (tableView: UITableView, 
                                     indexPath: IndexPath,
                                     outlineItem: SKHOutlineItem<DTOIndicatorSetting>
            ) -> IndicatorSettingCell? in
                
                let identifier = IndicatorSettingCellID.search.rawValue
                
                guard let cell = tableView.dequeueReusableCell(
                    withIdentifier: identifier,
                    for: indexPath
                ) as? IndicatorSettingCell else {
                    assertionFailure("IndicatorSettingCell 이 아니다 ")
                    return IndicatorSettingCell()
                }
                cell.selectionStyle = .none
                cell.data = outlineItem.contentItem
                return cell
        }
        dataSource?.defaultRowAnimation = .fade
    }
    
    func setupKeyboardObserver() {
        let nc = NotificationCenter.default
        showObserver = nc.addObserver(
            forName: UIResponder.keyboardWillShowNotification,
            object: nil,
            queue: OperationQueue.main
        ) { [weak self] notification in
            guard let self = self else { return }
            self.isShowKeyboard = true
            self.handleKeyboardNotification(notification)
        }
        
        hideObserver = nc.addObserver(
            forName: UIResponder.keyboardWillHideNotification,
            object: nil,
            queue: OperationQueue.main
        ) { [weak self] notification in
            guard let self = self else { return }
            self.isShowKeyboard = false
            self.handleKeyboardNotification(notification)
        }
    }
    
    // MARK: - Actions
    
    func handleKeyboardNotification(_ notification: Notification) {
        let previousNotificationName = self.previousNotificationName
        let previousKeyboardHeight = self.previousKeyboardHeight
        let name = notification.name
        self.previousNotificationName = name
        
        guard let userInfo = notification.userInfo,
              let keyboardFrame = userInfo[UIResponder.keyboardFrameEndUserInfoKey] as? CGRect
        else {
            return
        }
        
        let keyboardHeight = keyboardFrame.size.height
        self.previousKeyboardHeight = keyboardHeight
        if previousNotificationName == name && previousKeyboardHeight == keyboardHeight {
            return;
        }
        
        let duration = userInfo[UIResponder.keyboardAnimationDurationUserInfoKey] as? Double ?? 0.25
        let animationCurve = userInfo[UIResponder.keyboardAnimationCurveUserInfoKey] as? UInt ?? 7
        let options = UIView.AnimationOptions(rawValue: animationCurve)
        
        if name == UIResponder.keyboardWillHideNotification {
            let originalMaxOffsetY = tableView.skhMaxOffset().y - tableView.contentInset.bottom
            if originalMaxOffsetY < tableView.contentOffset.y {
                UIViewPropertyAnimator.runningPropertyAnimator(
                    withDuration: duration,
                    delay: 0.0,
                    options: options,
                    animations: {
                        self.tableView.contentOffset = CGPoint(x: 0.0, y: originalMaxOffsetY)
                        self.view.layoutIfNeeded()
                    }) { _ in
                        self.tableView.contentInset = .zero
                        self.tableView.scrollIndicatorInsets = .zero
                    }
            } else {
                tableView.contentInset = .zero
                tableView.scrollIndicatorInsets = .zero
            }
            return  //! return 해야한다.
        }
        
        if name == UIResponder.keyboardWillShowNotification {
            let bottomInset = keyboardHeight - view.safeAreaInsets.bottom
            tableView.contentInset = UIEdgeInsets(top: 0.0, left: 0.0, bottom: bottomInset, right: 0.0)
            tableView.scrollIndicatorInsets = UIEdgeInsets(top: 0.0, left: 0.0, bottom: bottomInset, right: 0.0)
        }
        
        guard let window = view.window else {
            return
        }
        
        let rect = view.convert(view.bounds, to: window)
        let upLength = rect.origin.y + rect.size.height + 4.0
        let total = upLength + keyboardHeight
        var movingOffset = tableView.contentOffset
           
        if total > window.bounds.height {
            let move = abs(window.bounds.height - total)
            movingOffset = CGPoint(x: movingOffset.x, y: movingOffset.y + move)
        }
                
        UIViewPropertyAnimator.runningPropertyAnimator(
            withDuration: duration,
            delay: 0.0,
            options: options,
            animations: {
                self.tableView.contentOffset = movingOffset
                self.view.layoutIfNeeded()
            },
            completion: nil
        )
    }
    
    private func updateEmptyStringModeForText(_ searchText: String) {
        if searchText.isEmpty {
            emptyStringMode = true
        } else {
            emptyStringMode = false
        }
    }
}

extension IndicatorSettingSearchController: UISearchBarDelegate {
    func searchBarShouldBeginEditing(_ searchBar: UISearchBar) -> Bool {
        updateEmptyStringModeForText(searchBar.text ?? "")
        return true
    }
    
    func searchBar(_ searchBar: UISearchBar, textDidChange searchText: String) {
        updateEmptyStringModeForText(searchText)
        if emptyStringMode == false {
            if isDebounceEnabled {
                debounceClosure(0.2) { [weak self] in
                    self?.updateSearchResult(searchBar, currentText: searchText)
                }
            } else {
                updateSearchResult(searchBar, currentText: searchText)
            }
        }
    }
    
    // MARK: - private Helper
    
    /// 디바운스(debounce) 효과가 부담스럽다면 디바운스만 제거하고 실행 Closure를 그냥 실행하면된다
    private func updateSearchResult(_ searchBar: UISearchBar, currentText: String) {
        guard let viewModel = viewModel,
              let dataSource = dataSource
        else {
            return
        }
        let snp = viewModel.snapshotForCurrentText(currentText)
        if snp.numberOfItems == 0 {
            self.noSearchResultView.searchText = currentText
            self.noSearchResultView.isHidden = false
            dataSource.apply(NSDiffableDataSourceSnapshot(), animatingDifferences: false)
        } else {
            self.noSearchResultView.isHidden = true
            self.tableView.contentOffset = .zero
            dataSource.apply(snp, animatingDifferences: false)
        }
    }
}

extension IndicatorSettingSearchController: UITableViewDataSource {
    func numberOfSections(in tableView: UITableView) -> Int {
        return 0
    }
    
    func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        return 0
    }
    
    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        return UITableViewCell()
    }
}

extension IndicatorSettingSearchController: UITableViewDelegate {
    func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
        if let outlineItem = dataSource?.itemIdentifier(for: indexPath) {
            selectItemCompletion?(outlineItem)
        }
    }
}

extension IndicatorSettingSearchController: UIGestureRecognizerDelegate {
    func gestureRecognizerShouldBegin(_ gestureRecognizer: UIGestureRecognizer) -> Bool {
        let location = gestureRecognizer.location(in: tableView)
        if tableView.indexPathForRow(at: location) != nil && !emptyStringMode {
            return false
        } else {
            return true
        }
    }
    
    // MARK: - private Helper
    
    @objc func endSearchMode(_ sender: UIGestureRecognizer) {
        view.window?.endEditing(true)
    }
}
